waku-bindings/src/node/mod.rs

```diff
@@ -82,100 +89,125 @@ impl WakuNodeHandle<Initialized> {
+    pub async fn relay_unsubscribe(&self, pubsub_topic: &PubsubTopic) -> Result<()> {
+        relay::waku_relay_unsubscribe(&self.ctx, pubsub_topic).await
     }
 
-    pub fn filter_subscribe(
+    pub async fn filter_subscribe(
         &self,
         pubsub_topic: &PubsubTopic,
         content_topics: Vec<WakuContentTopic>,
     ) -> Result<()> {
-        filter::waku_filter_subscribe(&self.ctx, pubsub_topic, content_topics)
+        filter::waku_filter_subscribe(&self.ctx, pubsub_topic, content_topics).await
     }
 
-    pub fn filter_unsubscribe(
+    pub async fn filter_unsubscribe(
         &self,
         pubsub_topic: &PubsubTopic,
         content_topics: Vec<WakuContentTopic>,
     ) -> Result<()> {
-        filter::waku_filter_unsubscribe(&self.ctx, pubsub_topic, content_topics)
+        filter::waku_filter_unsubscribe(&self.ctx, pubsub_topic, content_topics).await
     }
 
-    pub fn filter_unsubscribe_all(&self) -> Result<()> {
-        filter::waku_filter_unsubscribe_all(&self.ctx)
+    pub async fn filter_unsubscribe_all(&self) -> Result<()> {
+        filter::waku_filter_unsubscribe_all(&self.ctx).await
     }
 
-    pub fn lightpush_publish_message(
+    pub async fn lightpush_publish_message(
         &self,
         message: &WakuMessage,
         pubsub_topic: &PubsubTopic,
     ) -> Result<MessageHash> {
-        lightpush::waku_lightpush_publish_message(&self.ctx, message, pubsub_topic)
+        lightpush::waku_lightpush_publish_message(&self.ctx, message, pubsub_topic).await
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    pub async fn store_query(
+        &self,
+        pubsub_topic: Option<PubsubTopic>,
+        content_topics: Vec<WakuContentTopic>,
+        peer_addr: &str,
+        include_data: bool, // is true, resp contains payload, etc. Only msg_hashes otherwise
+        time_start: Option<u64>, // unix time nanoseconds
+        time_end: Option<u64>, // unix time nanoseconds
+        timeout_millis: Option<i32>,
+    ) -> Result<Vec<StoreWakuMessageResponse>> {
+        let mut cursor: Option<MessageHash> = None;
+
+        let mut messages: Vec<StoreWakuMessageResponse> = Vec::new();
+
+        loop {
+            let query = StoreQueryRequest::new()
+                .with_pubsub_topic(pubsub_topic.clone())
+                .with_content_topics(content_topics.clone())
+                .with_include_data(include_data)
+                .with_time_start(time_start)
+                .with_time_end(time_end)
+                .with_pagination_cursor(cursor)
+                .with_pagination_forward(true);
+
+            let response =
+                store::waku_store_query(&self.ctx, query, peer_addr, timeout_millis).await?;
+
+            messages.extend(response.messages);
+
+            if response.pagination_cursor.is_none() {
+                break;
+            }
+            cursor = response.pagination_cursor;
+        }
+
+        messages.reverse();
+
+        Ok(messages)
```

Comment:
It makes no sense too much for this method to be asynchronous as it stands. I would return a stream of values  (requested asynchronously) instead so it can be consumed later by the caller if needed. Otherwise, you keep the asynchronous API without allocation into a `Vec`.